File: 2. Linked List/C++/12. Palindrome_linked_list(vector).cpp
/*
 * Author : Rohit Shakya
 * Date   : May-2020
 * Compiler : g++ 4.9.2
 * flags    : -std=c++14
 * time complexity : O(n)
 * title : palindrome in linked list
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        vector<int> v;
        while(head!=NULL)
        {
            v.push_back(head->val);
            head=head->next;
        }
        int flag=0;
        for(int i=0;i<v.size();i++)
        {
            if(v[i]!=v[v.size()-i-1])
            {
                flag=1;
            }
        }
        return flag==1? false: true;
        
    }
};

File: 2. Linked List/Java/12. Palindrome_linked_list(vector).java
class Solution {
    public boolean isPalindrome(ListNode head) {
        return isNodesEqual(head, reverseHalfNode(findMiddleNode(head)));
    }
    
    private ListNode findMiddleNode(ListNode head) {
        ListNode slow = head, fast = head;
        while(fast != null) {
            slow = slow.next;
            fast = fast.next;
            if (fast != null) {
                fast = fast.next;
            }
        }
        return slow;
    }
    
    private ListNode reverseHalfNode(ListNode head) {
        ListNode rev = null, cur = head, prev = head;
        while(cur != null) {
            cur = cur.next;
            prev.next = rev;
            rev = prev;
            prev = cur;
        }
        return rev;
    }
    
    private boolean isNodesEqual(ListNode head, ListNode reversed) {
        while(reversed != null) {
            if (head.val != reversed.val) {
                return false;
            }
            head = head.next;
            reversed = reversed.next;
        }
        return true;
    }
}
